timeseries/src/autogluon/timeseries/models/abstract/tunable.py

```diff
@@ -0,0 +1,181 @@
+
+        try:
+            hpo_executor.validate_search_space(search_space, self.name)
+        except EmptySearchSpace:
+            return skip_hpo(self, train_data, val_data, time_limit=hpo_executor.time_limit)
+
+        train_path, val_path = self._save_with_data(train_data, val_data)
+
+        train_fn_kwargs = self._get_hpo_train_fn_kwargs(
+            model_cls=self.__class__,
+            init_params=self.get_params(),
+            time_start=time_start,
+            time_limit=hpo_executor.time_limit,
+            fit_kwargs=dict(
+                val_splitter=val_splitter,
+                refit_every_n_windows=refit_every_n_windows,
+            ),
+            train_path=train_path,
+            val_path=val_path,
+            hpo_executor=hpo_executor,
+        )
+
+        minimum_resources = self.get_minimum_resources(is_gpu_available=self._is_gpu_available())
+        hpo_context = disable_stdout if isinstance(hpo_executor, RayHpoExecutor) else nullcontext
+
+        minimum_cpu_per_trial = minimum_resources.get("num_cpus", 1)
+        if not isinstance(minimum_cpu_per_trial, int):
+            logger.warning(
+                f"Minimum number of CPUs per trial for {self.name} is not an integer. "
+                f"Setting to 1. Minimum number of CPUs per trial: {minimum_cpu_per_trial}"
+            )
+            minimum_cpu_per_trial = 1
+
+        with hpo_context(), warning_filter():  # prevent Ray from outputting its results to stdout with print
+            hpo_executor.execute(
+                model_trial=model_trial,
+                train_fn_kwargs=train_fn_kwargs,
+                directory=self.path,
+                minimum_cpu_per_trial=minimum_cpu_per_trial,
+                minimum_gpu_per_trial=minimum_resources.get("num_gpus", 0),
+                model_estimate_memory_usage=None,  # type: ignore
+                adapter_type="timeseries",
+            )
+
+            assert self.path_root is not None
+            hpo_models, analysis = hpo_executor.get_hpo_results(
+                model_name=self.name,
+                model_path_root=self.path_root,
+                time_start=time_start,
+            )
+
+        return hpo_models, analysis
+
+    def _get_default_hpo_executor(self) -> HpoExecutor:
+        backend = (
+            self._get_model_base()._get_hpo_backend()
+        )  # If ensemble, will use the base model to determine backend
+        if backend == RAY_BACKEND:
+            try:
+                try_import_ray()
+            except Exception as e:
+                warning_msg = f"Will use custom hpo logic because ray import failed. Reason: {str(e)}"
+                dup_filter.attach_filter_targets(warning_msg)
+                logger.warning(warning_msg)
+                backend = CUSTOM_BACKEND
+        hpo_executor = HpoExecutorFactory.get_hpo_executor(backend)()  # type: ignore
+        return hpo_executor
+
+    def _get_model_base(self) -> Self:
+        return self
```

Comment:
This method appears both in `TimeSeriesModelBase` and `TimeSeriesTunable`. Is this intended?

Reply:
fixed, thanks